trinity/sync/full/chain.py

```diff
@@ -238,7 +235,10 @@ def delayed_run_func(self, func: Callable[[], None], delay: float) -> None:
                 # peer returned no results, wait a while before trying again
                 delay = self.EMPTY_PEER_RESPONSE_PENALTY
                 self.logger.debug("Pausing %s for %.1fs, for sending 0 block bodies", peer, delay)
-                self.delayed_run_func(lambda: self._body_peers.put_nowait(peer), delay)
+                loop = self.get_event_loop()
+                loop.call_later(delay, partial(self._body_peers.put_nowait, peer))
```

Comment:
I think that we need to wrap `call_later` in a `BaseService.call_later` which retains the return value which I believe is the only way we can issue a cancellation when triggered.